File: lib/api/inventory.ts
import { apiClient, type ApiResponse } from "./index"
import type { InventoryItem, StockMovement } from "./types"

export class InventoryService {
  async getInventory(params?: {
    skip?: number
    limit?: number
    search?: string
    low_stock_only?: boolean
  }): Promise<ApiResponse<InventoryItem[]>> {
    const searchParams = new URLSearchParams()

    if (params?.skip) searchParams.append("skip", params.skip.toString())
    if (params?.limit) searchParams.append("limit", params.limit.toString())
    if (params?.search) searchParams.append("search", params.search)
    if (params?.low_stock_only) searchParams.append("low_stock_only", "true")

    const queryString = searchParams.toString()
    const endpoint = queryString ? `/inventory?${queryString}` : "/inventory"

    return apiClient.get<InventoryItem[]>(endpoint)
  }

  async updateInventoryItem(
    id: string,
    data: {
      current_stock?: number
      minimum_stock?: number
      maximum_stock?: number
      reorder_point?: number
      unit_cost?: number
    },
  ): Promise<ApiResponse<InventoryItem>> {
    return apiClient.put<InventoryItem>(`/inventory/${id}`, data)
  }

  async createStockMovement(movement: {
    product_id: string
    transaction_type: "purchase" | "sale" | "adjustment" | "return" | "transfer"
    quantity: number
    unit_cost?: number
    reference_id?: string
    notes?: string
  }): Promise<ApiResponse<StockMovement>> {
    return apiClient.post<StockMovement>("/inventory/movements", movement)
  }

  async getStockMovements(params?: {
    skip?: number
    limit?: number
    product_id?: string
    transaction_type?: string
  }): Promise<ApiResponse<StockMovement[]>> {
    const searchParams = new URLSearchParams()

    if (params?.skip) searchParams.append("skip", params.skip.toString())
    if (params?.limit) searchParams.append("limit", params.limit.toString())
    if (params?.product_id) searchParams.append("product_id", params.product_id)
    if (params?.transaction_type) searchParams.append("transaction_type", params.transaction_type)

    const queryString = searchParams.toString()
    const endpoint = queryString ? `/inventory/movements?${queryString}` : "/inventory/movements"

    return apiClient.get<StockMovement[]>(endpoint)
  }

  async getAnalytics(): Promise<
    ApiResponse<{
      total_products: number
      total_value: number
      low_stock_items: number
      out_of_stock_items: number
      recent_movements: number
      top_moving_products: Array<{ product_name: string; movements: number }>
    }>
  > {
    return apiClient.get("/inventory/analytics")
  }
}

export const inventoryService = new InventoryService()

File: backend/models/inventory.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum

class TransactionType(str, Enum):
    PURCHASE = "purchase"
    SALE = "sale"
    ADJUSTMENT = "adjustment"
    RETURN = "return"
    TRANSFER = "transfer"

class StockMovementCreate(BaseModel):
    product_id: str
    transaction_type: TransactionType
    quantity: int
    unit_cost: Optional[float] = None
    reference_id: Optional[str] = None  # Order ID, PO ID, etc.
    notes: Optional[str] = None

class StockMovementResponse(BaseModel):
    id: str
    product_id: str
    product_name: Optional[str] = None
    transaction_type: TransactionType
    quantity: int
    unit_cost: Optional[float] = None
    reference_id: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime
    created_by: str

class InventoryItemUpdate(BaseModel):
    current_stock: Optional[int] = Field(None, ge=0)
    minimum_stock: Optional[int] = Field(None, ge=0)
    maximum_stock: Optional[int] = Field(None, ge=0)
    reorder_point: Optional[int] = Field(None, ge=0)
    unit_cost: Optional[float] = Field(None, gt=0)

class InventoryItemResponse(BaseModel):
    id: str
    product_id: str
    product_name: Optional[str] = None
    product_sku: Optional[str] = None
    current_stock: int
    minimum_stock: int
    maximum_stock: int
    reorder_point: int
    unit_cost: float
    total_value: float
    last_updated: datetime
    status: str  # "in_stock", "low_stock", "out_of_stock"

class InventoryAnalytics(BaseModel):
    total_products: int
    total_value: float
    low_stock_items: int
    out_of_stock_items: int
    recent_movements: int
    top_moving_products: List[dict]
